youtube_dl/extractor/zingmp3.py

```diff
@@ -36,61 +65,100 @@ def _extract_item(self, item, fatal):
-                })
+                if v != 'VIP':
+                    formats.append({
+                        'ext': 'mp3',
+                        'format_id': k,
+                        'tbr': int_or_none(k),
+                        'url': self._proto_relative_url(v),
+                        'vcodec': 'none',
+                    })
         if not formats:
             if not fatal:
                 return
-            msg = item['msg']
+            msg = item.get('msg')
             if msg == 'Sorry, this content is not available in your country.':
                 self.raise_geo_restricted(countries=self._GEO_COUNTRIES)
-            raise ExtractorError(msg, expected=True)
         self._sort_formats(formats)
 
-        subtitles = None
         lyric = item.get('lyric')
-        if lyric:
-            subtitles = {
-                'origin': [{
-                    'url': lyric,
-                }],
-            }
+        if not lyric:
+            api = self.get_api_with_signature(name_api=self._SLUG_API.get("lyric"), param={
+                'ctime': self._TIMESTAMP,
+                'id': item_id})
+            info_lyric = self.download_json(api, item_id)
+            lyric = try_get(info_lyric, lambda x: x['data']['file'])
+        subtitles = {
+            'origin': [{
+                'url': lyric,
+            }],
+        } if lyric else None
 
         album = item.get('album') or {}
 
         return {
             'id': item_id,
             'title': title,
             'formats': formats,
-            'thumbnail': item.get('thumbnail'),
+            'thumbnail': item.get('thumbnail') or item.get('thumbnailM'),
             'subtitles': subtitles,
             'duration': int_or_none(item.get('duration')),
             'track': title,
-            'artist': item.get('artists_names'),
+            'artist': item.get('artistsNames') or item.get('artists_names'),
             'album': album.get('name') or album.get('title'),
-            'album_artist': album.get('artists_names'),
+            'album_artist': album.get('artistsNames') or album.get('artists_names'),
         }
 
     def _real_extract(self, url):
-        page_id = self._match_id(url)
-        webpage = self._download_webpage(
-            url.replace('://zingmp3.vn/', '://mp3.zing.vn/'),
-            page_id, query={'play_song': 1})
-        data_path = self._search_regex(
-            r'data-xml="([^"]+)', webpage, 'data path')
-        return self._process_data(self._download_json(
-            'https://mp3.zing.vn/xhr' + data_path, page_id)['data'])
+        mobj = re.match(self._VALID_URL, url)
+
+        song_id = mobj.group('id')
+        type_url = mobj.group('type')
+
+        api = self.get_api_with_signature(name_api=self._SLUG_API[type_url], param={
+            'ctime': self._TIMESTAMP,
+            'id': song_id})
+        info = self.download_json(api, song_id)
+        return self._process_data(info.get('data'), song_id, type_url)
```

Comment:
`_process_data(x,...)` expects `x` to be a `dict`. The downloaded 'JSON' may turn out to be a list or just `False`. So:
```
        info = self.download_json(api, song_id)
        info_data = try_get(info, lambda x: x['data'], dict)
        if not info_data:
            # if the site might often do this, not just when the API changes, add
            # `, expected=True`
            raise ExtractorError('API returned empty or invalid song data')
        return self._process_data(info_data, song_id, type_url)
```

Reply:
Updated.